File: src/main/java/io/github/romeoahmed/cursedoath/technique/RedBlast.java
package io.github.romeoahmed.cursedoath.technique;

import io.github.romeoahmed.cursedoath.domain.DomainInteractions;
import io.github.romeoahmed.cursedoath.world.LoadedChunks;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.ai.attributes.Attributes;
import net.minecraft.world.level.ServerExplosion;
import net.minecraft.world.phys.AABB;
import net.minecraft.world.phys.Vec3;

/// Damage and repulsion use the cover present at impact; excavation has its own budget.
final class RedBlast {
    private static final double RADIUS = TechniqueTuning.RED_RADIUS, KNOCKBACK = 4.5, LIFT = 0.8;
    private static final float DAMAGE = TechniqueTuning.RED_DAMAGE;
    // ServerExplosion offsets exposure samples toward +X/+Z by less than half a block.
    private static final double COVER_SAMPLE_MARGIN = 0.5;

    private RedBlast() {}

    static void impact(ServerPlayer player, Vec3 center, Vec3 direction) {
        var bounds = AABB.ofSize(center, RADIUS * 2, RADIUS * 2, RADIUS * 2);
        var level = player.level();
        for (var target : level.getEntitiesOfClass(LivingEntity.class, bounds)) {
            if (!TechniqueCombat.isActive(player, level)) return;
            if (!target.level().equals(level)) continue;
            repel(player, target, center, direction);
        }
    }

    private static void repel(ServerPlayer player, LivingEntity target, Vec3 center, Vec3 direction) {
        if (!TechniqueCombat.canAffect(player, target) || TechniqueCombat.hasInfinity(target)) return;
        if (DomainInteractions.contact(player, center, target.getBoundingBox().getCenter()) != null) return;
        double distance = Math.sqrt(target.getBoundingBox().distanceToSqr(center));
        var coverBounds = target.getBoundingBox().expandTowards(COVER_SAMPLE_MARGIN, 0, COVER_SAMPLE_MARGIN);
        if (distance >= RADIUS || !LoadedChunks.contains(player.level(), new AABB(center, center).minmax(coverBounds)))
            return;
        double strength = (1 - distance / RADIUS) * ServerExplosion.getSeenPercent(center, target);
        if (strength <= 0) return;
        var level = player.level();
        var source = level.damageSources().playerAttack(player);
        if (target.hurtServer(level, source, (float) (DAMAGE * strength))
                && !target.isRemoved()
                && target.level().equals(level)) {
            // Native knockback is horizontal; this technique also works when aimed vertically.
            double resistance = target.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE);
            TechniqueCombat.push(
                    target, direction.scale(KNOCKBACK).add(0, LIFT, 0).scale(strength * (1 - resistance)));
        }
    }
}

File: src/main/kotlin/io/github/romeoahmed/cursedoath/technique/RedBlast.kt
package io.github.romeoahmed.cursedoath.technique

import io.github.romeoahmed.cursedoath.world.hasLoadedChunks
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.ai.attributes.Attributes
import net.minecraft.world.level.ServerExplosion
import net.minecraft.world.phys.AABB
import net.minecraft.world.phys.Vec3
import kotlin.math.sqrt

/** Damage and repulsion use the cover present at impact; excavation follows under its own budget. */
internal object RedBlast {
    private const val RADIUS = TechniqueTuning.RED_RADIUS
    private const val DAMAGE = TechniqueTuning.RED_DAMAGE
    private const val KNOCKBACK = 4.5
    private const val LIFT = 0.8

    // ServerExplosion offsets exposure samples toward +X/+Z by less than half a block.
    private const val COVER_SAMPLE_MARGIN = 0.5

    fun impact(
        player: ServerPlayer,
        center: Vec3,
        direction: Vec3,
    ) {
        val level = player.level()
        val bounds = AABB.ofSize(center, RADIUS * 2, RADIUS * 2, RADIUS * 2)
        for (target in level.getEntitiesOfClass(LivingEntity::class.java, bounds)) {
            repel(player, target, center, direction)
        }
    }

    private fun repel(
        player: ServerPlayer,
        target: LivingEntity,
        center: Vec3,
        direction: Vec3,
    ) {
        if (!TechniqueCombat.canAffect(player, target) || TechniqueCombat.hasInfinity(target)) return
        val distance = sqrt(target.boundingBox.distanceToSqr(center))
        val coverBounds = target.boundingBox.expandTowards(COVER_SAMPLE_MARGIN, 0.0, COVER_SAMPLE_MARGIN)
        if (distance >= RADIUS ||
            !player.level().hasLoadedChunks(AABB(center, center).minmax(coverBounds))
        ) {
            return
        }
        val strength = (1 - distance / RADIUS) * ServerExplosion.getSeenPercent(center, target)
        if (strength <= 0) return
        val damage = (DAMAGE * strength).toFloat()
        val level = player.level()
        val source = level.damageSources().playerAttack(player)
        if (target.hurtServer(level, source, damage)) {
            // LivingEntity.knockback is horizontal; the technique must also work when aimed up or down.
            val resistance = target.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE)
            val force =
                direction
                    .scale(KNOCKBACK)
                    .add(0.0, LIFT, 0.0)
                    .scale(strength * (1 - resistance))
            TechniqueCombat.push(target, force)
        }
    }
}
